General payu docs for Run a Model section

ccarouge opened the pull request:
# ACCESS-Hive Docs

Thank you for submitting a pull request to the ACCESS-Hive Docs Project. 🎉

More assistance is available in the [how to contribute section on ACCESS-Hive Docs](https://docs.access-hive.org.au/about/contribute/)

## Description

Write a general payu doc page for the Run a Model section. 

Fixes #978 

❓ The issue #978 includes the issue #934 on hosting payu experiments on GitHub. However, the discussion on that issue does not seem to be finalised as to where to include this information. @atteggiani seems to prefer a tutorial page, which would not be the general payu page included in this PR. So for now, this PR does not address issue #934. ❓ 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue, e.g. dead links)
- [x] New link / content

## Checklist:

- [ ] The new content is accessible and located in the appropriate section
- [ ] My changes do not break navigation and do not generate new warnings
- [ ] I have checked that the links are valid and point to the intended content
- [ ] I have checked my code/text and corrected any misspellings

When your pull request is ready please [request a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review). 

Unless there is a specific person you want your PR to be reviewd by, please select the **Hive Docs Team**: `ACCESS-NRI/hivedocsteam`. This ensures the load for reviewing pull requests is shared equitably.

ccarouge committed:
Start general payu docs

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,406 @@
+<terminal-window>
+    <terminal-line data="input">payu --version</terminal-line>
+    <terminal-line lineDelay="1000">1.1.3</terminal-line>
+</terminal-window>
```
~I would place the current latest version (`1.2.0`) but wouldn't worry about keeping it up to date in the future.~

Actualy, thinking it through, I would remove this animation. 
I don't think it really provides help for a simple command like `payu --help`

ccarouge commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,406 @@
+
+    payu run -n 10
+
+## Monitor the experiment
```
I feel all the info here and the subsection "Stop the run" is valid for any job on Gadi and not just payu. Should this be somewhere else?

aidanheerdegen replied:
Probably. But if no such place exists make an issue and move on would be my approach.

ccarouge commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,406 @@
+</terminal-window>
+
+This can help to isolate issues such as permissions problems accessing files and directories, missing files or malformed/incorrect paths.
+
```
The documentation of ACCESS-OM2 has a [section on Model Live Diagnostics](https://docs.access-hive.org.au/pr-previews/1107/models/run_a_model/run_access-om2/#model-live-diagnostics). That isn't about payu so I haven't included it but it might be worth adding since it helps with the monitoring?

ccarouge commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,406 @@
+As seen previously in [Run an experiment](#run-an-experiment), _payu_ handles this by splitting a simulation in sections and automatically running each section one after the other in separate jobs. As such the total run length is the product of:
+
+- the run length of each individual submission. This is set differently depending on the model, so refer to the [Run a Model][Run a Model] page for details. It is important to set this run length so that the simulation finishes within the maximum wall time for the job.
+- the number of automatic resubmissions by _payu_. This is set through the command line option `-n`. It defaults to 1 and there is no limit on this number.
```
This seems like a straight repeat from the section "Run an experiment". Is it worth repeating? I think it is worth having a section about changing the run length because people would be more likely to find the info here than under a section named "Run an experiment". We could rename "Run an experiment" section to make the title more descriptive or leave the repeat (so that the section about modifying the config also talks about changing the length of the simulation). Not sure what is best

jo-basevi replied:
I think it doesn't hurt having the repeat of information here, but I also not have any strong thoughts on it. It is worth specifying that the maximum wall time  for the individual job is set by the [`walltime:` entry] in the `config.yaml`?

ccarouge commented on docs/models/run_a_model/payu.md:
````diff
@@ -0,0 +1,406 @@
+project: ol01
+```
+
+For model configurations and output to be saved to a `/scratch` storage allocation other than `project` (or your default if `project` is not set) then also set `shortpath` to the desired path. 
````
Should the `shortpath` path exists before running payu or will payu create it? It is worth specifying here.

jo-basevi replied:
Yes, payu does create the `shortpath` directory if it doesn't exist

ccarouge commented on docs/models/run_a_model/payu.md:
````diff
@@ -0,0 +1,406 @@
+      - '*.nc.*'
+      - 'iceh.????-??-??.nc'
+```
+To enable syncing, change `enable` to `True`, and set `path` to a location on `/g/data`, where _payu_ will copy output and restart folders. A sensible `path` could be: `/g/data/$PROJECT/$USER/<model>/experiment_name/`.
````
Should this path exists before running payu or will it be created by payu? Worth specifying here as well.

ccarouge commented on docs/models/run_a_model/payu.md:
````diff
@@ -0,0 +1,406 @@
+input: /g/data/ik11/inputs/access-om2/input_20201102/common_1deg_jra55
+```
+
+The `name` field is not actually used for the configuration run, so it can be safely ignored.
````
Is that true of all models that `name` is not used or is it an OM2 thing?

jo-basevi replied:
I think `name` is used when there are submodel configurations. It's mostly for naming subdirectories in work/archive folders, e.g. "ocean", "ice". For the top-level model, I think 'name' is overwritten with the model name, e.g. `access-om2`.

ccarouge committed:
Changes from reading the rendered version

anton-seaice commented on mkdocs.yml:
You need to add payu.md to the run_a_model page as well (docs/models/run_a_model/index.md)

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,417 @@
+
+[_payu_](https://github.com/payu-org/payu) is a workflow manager tool for running numerical models in supercomputing environments. It is an open-source software, distributed under an Apache 2.0 Licence.
+
+You can find a [full documentation](https://payu.readthedocs.io/en/latest/) for _payu_ online. This page summarises the _payu_ capabilities that are most commonly required to run an ACCESS model's configuration.
```
```suggestion
This page summarises the _payu_ capabilities that are most commonly required to run an ACCESS model's configuration. There is also [technical documentation](https://payu.readthedocs.io/en/latest/) for how to configure _payu_. 
```

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,417 @@
+[_payu_](https://github.com/payu-org/payu) is a workflow manager tool for running numerical models in supercomputing environments. It is an open-source software, distributed under an Apache 2.0 Licence.
+
+You can find a [full documentation](https://payu.readthedocs.io/en/latest/) for _payu_ online. This page summarises the _payu_ capabilities that are most commonly required to run an ACCESS model's configuration.
+
```
Do we need to be explicit on what is covered in this page compared to the pages specific to each model.

anton-seaice replied:
This is reasonable well covered by the comment added above


> This page is to be used in conjunction with the [Run a Model][Run a Model] page for the chosen configuration ...

anton-seaice reviewed (commented):
I started reviewing , and now realised that this is Draft. Let me know when you want a review

ccarouge committed:
Changes following comments by @anton-seaice

jo-basevi commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,406 @@
+By default, every `restart_freq`, _payu_ removes intermediate restart files, keeping only: 
+- the two most recent restarts
+- restarts corresponding to the `restart_freq` interval
+For example, a `restart_freq` set to `1YS` would keep the restart files at the end of each model year, whereas `restart_freq` set to `5YS` would keep those at the end of every fifth model year.
```
It keeps the first restart from the start of the 5th year from the last permanently-saved restart. So if there was the following restarts:

```
restart00X: Last date in restart file
restart000: 01/01/2000
restart001: 01/01/2003
restart002: 01/01/2006
restart003: 01/01/2009
restart004: 01/01/2012
restart005: 01/01/2015
```

The saved restarts would be:

```
restart000: 01/01/2000
restart002: 01/01/2006 (first restart >= 01/01/2005)
restart004: 01/01/2012 (first restart >= 01/01/2011)
restart005: 01/01/2015 (keeps immediate restarts < 01/01/2017)
```

jo-basevi reviewed (commented):
Sorry for the late initial review! I've just added a couple comments so far. Overall it seems to read well to me

ccarouge committed:
(#978) - Minor clarifications

ccarouge committed:
(#978) - Incorporate Jo's feedback

ccarouge committed:
(#978) - Formatting.

ccarouge committed:
(#978) - Additions from the run ESM1.5 page.

ccarouge commented:
@ACCESS-NRI/hivedocsteam @jo-basevi @aidanheerdegen  We need a logo for payu for the Run a Model index page: https://docs.access-hive.org.au/pr-previews/1107/models/run_a_model/ (reusing rose-cylc image for now). Any idea?

ccarouge committed:
(#978) - try and improve layout on run a model index page.

ccarouge commented:
@anton-seaice If you need to for the review, I've started an rewrite of the ESM1.5 page to see what it would look like for the models using this payu page. See #1121

anton-seaice committed:
Merge remote-tracking branch 'origin/development' into 978-General-how-to-use-Payu-Docs

jo-basevi commented on docs/models/run_a_model/payu.md:
````diff
@@ -0,0 +1,495 @@
+        years: 1
+        months: 0
+        days: 0
+```
````
I think using `runtime:` dictionary in the `config.yaml` is ACCESS-ESM1.5 specific. There's different instructions for setting runtime for ACCESS-OM2 (https://docs.access-hive.org.au/models/run_a_model/run_access-om2/#change-run-length) and ACCESS-OM3 (https://docs.access-hive.org.au/models/run_a_model/run_access-om3/#change-run-length-and-restart-period)

jo-basevi commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+Now some practical examples:
+
+- **Run 20 years of simulation with resubmission every 5 years**<br>
+    To have a _total experiment length_ of 20 years with a 5-year resubmission cycle, leave [`runtime`](#runtime) as the default value of `1 year`, set `runspersub` to `5` and increase `walltime` to allow for 5 years of simulation within a PBS job. Then, run the configuration with `-n` set to `20`:
```
The default value for runtime might be model configuration dependent. E.g. the OM2 `1deg_jra55_ryf` configuration is 5 years, and the higher resolutions are  shorter -e.g. `01deg_jra55_ryf` is 3 months.

jo-basevi reviewed (commented):
Thanks @ccarouge, I think all my earlier comments were resolved.

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+
+Once you have chosen the configuration, you need to:
+
+- identify the `<repository>` and `<branch>` name the configuration is stored under on GitHub.
```
Are repository and branch defined somewhere ? maybe in the glossary ?

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+
+For example, say you want to do a sensitivity experiment to the diffusivity in ACCESS-OM2 using the configuration `release-1deg_jra55_ryf`. You decide to:
+
+- `<repository>` and `<branch>`: base your experiment off the branch, `release-1deg_jra55_ryf`, from the repository, https://github.com/ACCESS-NRI/access-om2-configs
```
```suggestion
- `<repository>` and `<branch>`: base your experiment off the branch, `release-1deg_jra55_ryf`, from the [ACCESS-OM2 configs](https://github.com/ACCESS-NRI/access-om2-configs) repository, 
```

for consistency with the rest of the hive docs

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+
+The general layout of a _payu_-supported model run consists of two main directories:
+
+- The _control_ directory contains the model configuration and serves as the execution directory for running the model. You created the _control_ directory when you cloned the configuration you want to use.
```
whats an execution directory?

Just say its the place the model is run from

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+- The _control_ directory contains the model configuration and serves as the execution directory for running the model. You created the _control_ directory when you cloned the configuration you want to use.
+- The _laboratory_ directory, where all the model components reside. It is typically `/scratch/$PROJECT/$USER/<model_name>` and is created by _payu_. `$PROJECT` and `$USER` are environment variables on _Gadi_ that points to your default project and your username respectively.
+
+This separates the small text configuration files from the larger binary outputs and inputs. In this way, the _control_ directory can be in the `$HOME` directory (as it is the only filesystem actively backed-up on _Gadi_). The quotas for `$HOME` are low and strict, which limits what can be stored there, so it is not suitable for larger files.
```
The connection between the "model configuration" being the "small text configuration files" and the "model components"/lab directory being the "larger binary outputs and inputs" is not clear

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+
+This separates the small text configuration files from the larger binary outputs and inputs. In this way, the _control_ directory can be in the `$HOME` directory (as it is the only filesystem actively backed-up on _Gadi_). The quotas for `$HOME` are low and strict, which limits what can be stored there, so it is not suitable for larger files.
+
+The _laboratory_ directory is a shared space for a user's _payu_ experiments using the same model. Inside the _laboratory_ directory there are two areas:
```
What does shared mean here ?

The start of this section talks about an individual model configuration in the singular, and now we say the _laboratory_ directory is actually shared across all experiments using this model 

I think i would stick to only talking about one configuration

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+- `work` &rarr; for temporary storage of files needed by the model while it runs. _payu_ creates and removes directories and files in this directory upon successful completion of runs.
+- `archive` &rarr; for storing the output following each successful run.
+
+Within each of the above directories, _payu_ automatically creates subdirectories uniquely named according to the experiment being run.<br>
```
This sentence feels out of place / unclear.
We should encourage users to use the symlinks in teh control directory rather than understand this level of detail

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+
+## Run the configuration
+
+_payu_ manages the experiment through a [PBS job][PBS job] that it self-submits.
```
```suggestion
_payu_ manages the experiment through a [PBS job][PBS job] that it submits.
```

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+Within each of the above directories, _payu_ automatically creates subdirectories uniquely named according to the experiment being run.<br>
+_Payu_ also creates symbolic links in the _control_ directory pointing to the `archive` and `work` directories.
+
+This design allows multiple self-resubmitting experiments that share common executables and input data to be run simultaneously.
```
I think its trying to say

"As the _work_ and _archive_ directories are unique to each experiment, this allows multiple experiments to be run simultanesouly"

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+## Outputs organisation
+
+At the end of a successful model run, output files, restart files and log files are moved from the `work` directory to the `archive` directory.<br>
+Symbolic links to these directories are also provided in the _control_ directory for convenience.
```
```suggestion
Symbolic links to these directories are provided in the _control_ directory.
```

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+
+## Edit a _payu_ configuration
+
+The modifications discussed in this section can change how the model and its components are configured, or the way the model is run by _payu_.
```
```suggestion
The modifications discussed in this section can change the way the model is run by _payu_.
```

(refer to model specific pages for how to chang eth model )

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+
+#### Submodels {: .no-toc }
+
+Coupled models deploy multiple submodels, a.k.a. the model components.
```
```suggestion
Some coupled models simultanouesly run multiple executables for each submodel, a.k.a. the model components.
```

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+For more information about specific `userscripts` fields, check the relevant section of [_payu_ Configuration Settings documentation](https://payu.readthedocs.io/en/latest/config.html#postprocessing).
+
+#### Postscripts {: .no-toc }
+Postprocessing scripts that run after _payu_ has completed all steps of each run (for example, with `payu run -n 10`, the postscript will run 10 times). Scripts that might alter the output directory, for example, can be run as postscripts. These run in PBS jobs separate from the main model simulation.
```
```suggestion
Postprocessing script that runs after _payu_ has completed all steps of each run (for example, with `payu run -n 10`, the postscript will run 10 times). Scripts that might alter the output directory, for example, can be run as as a postscript. These run in PBS jobs separate from the main model simulation.
```

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,406 @@
+These options are typically useful to modify the physics used in the model, the input data, or the model variables saved in the output files.
+
+These configuration options are specified in files located inside a subfolder of the _control_ directory, named according to the submodel's `name` specified in the `config.yaml` `submodels` section (e.g., configuration options for the _ocean_ component are in the `ocean` sub-directory).<br>
+To modify these options please refer to the User Guide of the respective model component.
```
Yeah config docs :)

anton-seaice reviewed (commented):
Thanks @ccarouge

atteggiani commented:
Sorry to chime in on this. I haven't followed the whole discussion and haven't reviewed this page yet, but I think it's great that discussions are happening and this page gets a lot of inputs! 

I just wanted to give my opinion on the "trial and error" logic here. 

> Many users will figure out how work/archive etc works through trial and error, so the focus should be on the information to get users started

I personally don't think this should be our main goal. While it’s true that we can’t explain everything and some trial and error is inevitable (that's probably how everyone learns in the first place), I think we should aim to make the information on our documentation pages as clear as possible, rather than relying on users to figure out things out through trial/error.

ccarouge commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+
+All model configurations are hosted on GitHub.<br>
+
+The first step is to choose a configuration from those available and identify the branch name for that configuration, following information on the [Run a Model][Run a Model] page of your chosen model.<br>
```
"and identify the branch name for that configuration": doesn't this mean a configuration is a branch?

anton-seaice replied:
I suggests a sentence above

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,495 @@
+    
+## Get the model configuration
+
+All model configurations are hosted on GitHub.<br>
```
```suggestion
All model configurations are hosted on GitHub. There is one configuration repository for each model. Individual branches in this repository represent individual configurations.<br>
```

aidanheerdegen commented:
> @aidanheerdegen Reading the consultation, it means we don't yet have a logo, but it doesn't mean we can't design one. Or am I misunderstanding your conversation with Marshall?

Yes you are correct. I think he would be happy for us to make something.

ccarouge committed:
(#560) - Redesign after review.

ccarouge committed:
(#978) - Additional changes

ccarouge committed:
(#978) - A few more edits

ccarouge committed:
Update diagram for payu directory structure

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,476 @@
+
+A representation of the data organisation for _payu_ is given in the following diagram:
+
+![payu directory structure](/docs/assets/payu_directory_structure.png){: class="img-contain white-background round-edges with-padding" loading="lazy"}
```
You should not include the `docs` folder in the link: `/assets/payu_directory_structure.png`

ccarouge committed:
General how to use Payu Docs
Fixes #978

ccarouge committed:
Fixing image link

ccarouge commented:
@anton-seaice I've redesigned this page. I'm putting just you as a reviewer for now to see if you think it's better and worth bothering others for review.

I also let you mark your comments as resolved if you think I've addressed them appropriately.

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -9,15 +9,65 @@
-There is also [technical documentation](https://payu.readthedocs.io/en/latest/) for how to configure _payu_. 
+![payu directory structure](/assets/payu_directory_structure.png){: class="example-img" loading="lazy"}
+
+This design was chosen to separate the small files that define the configuration and the larger binary output and input files needed for a realisation of a configuration. This ensures the configuration definition is easy to back up and share. It also optimises the use of different filesystems on high-performance computers. Finally, this layout ensures several experiments that share common executa ...
```
I think I would put this paragraph before the diagram

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,473 @@
+The `archive` and `work` directories for an experiment are most easily accessed through the symbolic links created in the _control_ directory.
+
+!!! warning
+    Files on the `/scratch` drive, such as the _laboratory_ directory, might get deleted if not accessed for several days and the `/scratch` drive is limited in space. For these reasons, all model runs which are to be kept should be moved to `/g/data/` by enabling the `sync` step in _payu_. To know more refer to [Syncing output data](#syncing-output-data-to-long-term-storage).
```
```suggestion
    Files on the `/scratch` drive, such as the _laboratory_ directory, might get deleted if not accessed for several days. All experiments which are to be kept should be moved to `/g/data/` by enabling the `sync` step in _payu_. To know more refer to [Syncing output data](#syncing-output-data-to-long-term-storage).
```

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,473 @@
+
+- `<repository>` and `<branch>`: base your experiment off the branch, `release-1deg_jra55_ryf`, from the repository, `https://github.com/ACCESS-NRI/access-om2-configs`
+- `<configurations-directory>`: store the all your ACCESS-OM2 configurations under `~/access-om2/`
+- `<experiment-name>`: name your experiment `diffuse_test1-1deg_jra55_ryf`
```
this should be branch-name I think

anton-seaice reviewed (commented):
This looks great - thanks @ccarouge . From a first look I find it much clearer.

anton-seaice commented on docs/assets/payu_directory_structure.png:
I quickly looked at mermaid and it seemed overly restrictive. What did you make the diagram in ? Maybe we just have to include the source file too?

ccarouge replied:
That's the best I came up with in mermaid: https://mermaid.live/edit#pako:eNqtVktv4zYQ_isEg8VuAdvw24kOBTbZ7a3XHrrugZJGMhGK1PKRWBvkv3dI6mE7crYtah9sDme-4ffNcKQXmqkcaEJTobLHaQqW7WWmhKukIau9JCRsJJmSVivhDYR0-4u4tNwKWHz6tqetF8m5hswq3ezpX79EpxxMFnxyKLjklitJVEHsARBOFrx0mnnjhHCZCYfuCdljaM0ad-khWQWCG0sKLsDMZrMhzb1P8az042B68CamswN_gs4KMvc_pmYZDCQFS68TXHoY9FB4CtVcoxi8DNpZCaRQmjDiDOiPhsCxBs0rkNaQnFkW2DnDZYk-BhmRCishOv7eLpguQZOUS6YbtNcOY30YkzlRzvr1kD5yCNzD-i2PjsrqrUh9PB7T7yC9fmMMp0Vae6SBGdJ0kn93MCYP6TRaY4yFqkaRkFYnFjZDKGcgKAFyyEnahAYJwpDnA24Tbol20tPuUdtinvyJVLqa_0SNzUh_nAviN_-hINv_Isi27RkIbGNhsXmEUM--DYBlh6CKcVkGxhROeA3ek6D9uSfTKUHspkqV4BkRXGLN0fgr6frk4bpLr5_3-_CBVIxLkqoj1sjYBh1Lv5EJZswXKEiplatxR6tHSFLBssdJXEyfeW4PyaY-9v7Y2mkbMNi66RHsb7JGfcdTp7z0F6VAgKnhPyBZzetjl13hJee2SeZDpjix2rAL87I3twfQUKknaInj9c9Bm7PsUsVM5G3CATvMv9713L48s4fBhCQBy3eJOMHGUxrVddCe7ubrarddfCV4iWTpZ6lQTsfJM7R_hHu4BtdidMkR1XcHdkWsBD3VPEKF_b7UDkYrHV3jxDmpzdLXZuAxWqYYOsyiLtfN55X_djE5MwemNWuSzZzs5uenWNdHcnGMNWkJn8OMpA5zirx7ylaqVub31OpcOhJR7p9pthnX7CJ6XLZ2Yv1Psm3_lWy999WD0glqxXOaWI3K0go0XnJc0pfQghSnYIWjGF8AaMoM_puc2P9gmrMUJ4F3eIlTb09rnLf4NHnwmWPkTYGfNjQEg7b8fRev92-s4qKJ-0-gcybZCMh9GAMnUGHe7an3e93LV2RYM_mnUlVHEnujPNCkYMLgytX4CgBfOCs1q3qrxrHtUZ20NFntAgZNXuiRJrfb2Wq9W94tl4vF3e36bkIbmqzn29nmdUJ_hDzz2e1ufXf6mVDIOT5Xfo-veOFN7_Vv9kFTPQ

I've made the diagram with lucid chart. There is no source file.

ccarouge committed:
(#978) - edits for review comments

ccarouge committed:
Merge branch 'development' into 978-General-how-to-use-Payu-Docs

aidanheerdegen commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+The `archive` and `work` directories for an experiment are most easily accessed through the symbolic links created in the _control_ directory.
+
+!!! warning
+    Files on the `/scratch` drive, such as the _laboratory_ directory, might get deleted if not accessed for several days. All experiments which are to be kept should be moved to `/g/data/` by enabling the `sync` step in _payu_. To know more refer to [Syncing output data](#syncing-output-data-to-long-term-storage).
```
```suggestion
    Files on the `/scratch` drive, such as the _laboratory_ directory, might be deleted if not accessed for 90 days. All experiments which are to be kept should be moved to `/g/data/` by enabling the automated `sync` step in _payu_. To know more refer to [Syncing output data](#syncing-output-data-to-long-term-storage).
```

aidanheerdegen commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+    !!! warning
+        Different model configurations will likely require you to **join additional projects**. Please refer to the [Run a Model][Run a Model] page of your chosen configuration for the list of additional projects.
+
+## Payu setup
```
I am concerned this could be confusing as there is a `payu setup` command.

How about "Accessing payu"?

aidanheerdegen commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+    payu --version
+
+## Run an experiment
+
```
I so wish we had some control experiments we could point to at this point. I think control experiments are what most users would expect to base their climate experiments on, and the fork -> clone workflow is much much much easier to explain and understand.

ccarouge replied:
We should discuss this "control experiments are what most users would expect to base their climate experiments on". I'm not sure it is as true as you think. But not in this PR.

aidanheerdegen commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+- identify the `<repository>` and `<branch>` name the configuration is stored under on GitHub. See the information on the [Run a Model][Run a Model] page of your chosen model for this step.
+- decide where on Gadi to store all your _payu_ experiments, `<configurations-directory>`, typically a folder under $HOME. This directory must exist before running _payu_.
+- decide on a name for your experiment, `<local-branch>`. It is recommended to choose a descriptive name, specific to your experiment.
+- decide on a directory name to store the experiment, `<control-directory>` (created by _payu_). The `control` directory is a git repository. Experiments are saved as branches in this repository, making it possible to use the same `control` directory for several experiments. For this reason, we recommend to always set the `<local-branch>`. For more information refer to this [payu tutorial](https: ...
```
```suggestion
- decide on a directory name to store the experiment, `<control-directory>` (created by _payu_). The `control` directory is a git repository. Experiments are branches in this repository, making it possible to use the same `control` directory for several experiments. For more information refer to this [payu tutorial](https://forum.access-hive.org.au/t/access-om2-payu-tutorial/1750#select-experiment-12).
```

atteggiani replied:
```suggestion
- Choose a name of your [_control directory_](#control-directory) (automatically created by _payu_). The _control directory_ is a Git repository where experiments are stored as branches, allowing the same _control directory_ to be used for multiple experiments. For this reason, it is recommended always setting the `<experiment-name>`. For more information, refer to [this payu tutorial](https://forum.access-hive.org.au/t/access-om2-payu-tutorial/1750#select-experiment-12).
```

aidanheerdegen commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+    mkdir -p ~/access-om2/
+    cd ~/access-om2/
+    payu clone -b diffuse_test1-1deg_jra55_ryf -B release-1deg_jra55_ryf https://github.com/ACCESS-NRI/access-om2-configs diffuse_exps-1deg_jra55_ryf
+    cd diffuse_exps-1deg_jra55_ryf
```
```suggestion
    mkdir -p ~/experiments/access-om2/
    cd ~/experiments/access-om2/
    payu clone -b test1 -B release-1deg_jra55_ryf https://github.com/ACCESS-NRI/access-om2-configs 1deg_jra55_ryf-diffusivity
    cd 1deg_jra55_ryf-diffusivity
```
Just putting in the modified suggested names from above.

aidanheerdegen commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+
+Coupled models deploy multiple submodels, a.k.a. the model components.
+
+This section of the _payu_ configuration file specifies the submodels, the configuration options required to execute the model correctly and the location of all inputs required for this submodel.
```
```suggestion
This section of the _payu_ configuration file specifies the submodels, the configuration options required to execute the model component correctly and the location of all inputs required for this submodel.
```

aidanheerdegen commented on docs/models/run_a_model/payu.md:
````diff
@@ -0,0 +1,474 @@
+```yaml
+runlog: true
+```
+When running a new configuration, _payu_ automatically commits changes with _git_ if `runlog` is set to `true`.
````
```suggestion
When running an experiment,  if `runlog` is set to `true`, _payu_ saves a history of the experiment. It does this using _git_, by automatically committing changes to the control directory repository.
```

aidanheerdegen reviewed (changes requested):
I've suggested quite a bit ... hope it is helpful.

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+[PBS job]: https://opus.nci.org.au/display/Help/4.+PBS+Jobs
+[Run a Model]: /models/run_a_model
+
+# Run models using payu
```
```suggestion
# Run models using Payu
```

ccarouge replied:
payu is always lowercase in the text. Why would we put it uppercase on the page title? Or do you want to use Title Case: "Run Models Using Payu"?

atteggiani replied:
Yes, Title Case would make sense in my opinion. 
However, it is currently not consistent across other pages, so feel free to leave this all lower case.

ccarouge replied:
My tomorrow self will decide on what happens here 😄

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+
+### Configuration versus experiment
+
+These terms are not interchangeable although they are closely related.
```
```suggestion
The terms _configuration_ and _experiment_ are not interchangeable although they are closely related.
```

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+
+These terms are not interchangeable although they are closely related.
+
+A configuration defines a specific way to run the model it relates to. 
```
```suggestion
A *configuration* defines a specific way to run the model it relates to. 
```

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+- a given set of input files (ancillaries, forcings, restarts)
+- a given set of physical and modelling options for each model component used by this model
+
+Changing any one of these items creates a new configuration
```
```suggestion
Changing any one of these elements creates a new configuration.
```

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+### Data organisation and _payu_'s directories designation
+
+!!! info
+    `payu` will create all the directories it needs. They do not need to be created beforehand.
```
```suggestion
    _payu_ will create all the directories it needs. They do not need to be created beforehand.
```

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+To get a local copy of a configuration, you need to:
+
+- identify the `<repository>` and `<branch>` name the configuration is stored under on GitHub. See the information on the [Run a Model][Run a Model] page of your chosen model for this step.
+- decide where on Gadi to store all your _payu_ experiments, `<configurations-directory>`, typically a folder under $HOME. This directory must exist before running _payu_.
```
I might not have fully understood this sentence. 
Based on my understanding, I would rephrase to:
```suggestion
- Create your [_control directory_](#control-directory) on _Gadi_, where all the model configurations will be stored. We suggest creating it in your `$HOME` directory.
```

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+
+- identify the `<repository>` and `<branch>` name the configuration is stored under on GitHub. See the information on the [Run a Model][Run a Model] page of your chosen model for this step.
+- decide where on Gadi to store all your _payu_ experiments, `<configurations-directory>`, typically a folder under $HOME. This directory must exist before running _payu_.
+- decide on a name for your experiment, `<local-branch>`. It is recommended to choose a descriptive name, specific to your experiment.
```
```suggestion
- Choose a name for your experiment. This experiment name will be referred to as `<local-branch>`. It is recommended to choose a descriptive name, specific to your experiment.
```

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+    <terminal-line linedelay=0 class="keep-blanks">&lt;job-ID&gt;.gadi-pbs      &lt;job-name&gt; &lt;\$USER&gt;           &lt;time&gt;   R normal-exec</terminal-line>
+</terminal-window>
+
+The default name of your job is the name of the _payu_ _control_ directory.<br>
```
```suggestion
The default name of your job is the name of the _payu_ [_control_ directory](#control-directory).<br>
```

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+#### Model log files {: .no-toc }
+
+While the model is running, the standard output and error streams are saved in the _control_ directory. Refer to the [Run a Model][Run a Model] page for the model you are using for the list of logging filenames for your model.<br>
+You can examine the contents of these files to check on the status of a run as it progresses (or after a failed run has completed).
```
```suggestion
While the model is running, the standard output and error streams are saved in the _control_ directory. Refer to the [Run a Model][Run a Model] page for your chosen model for its list of logging filenames.<br>
You can examine the contents of these log files to check on the status of a run as it progresses (or after a failed run has completed).
```

atteggiani commented on docs/models/run_a_model/payu.md:
````diff
@@ -0,0 +1,474 @@
+restart_freq: '50YS'
+```
+
+The most recent sequential restarts are retained, and only deleted after a permanently archived restart file has been produced.
````
```suggestion
The most recent sequential restarts are retained, and only deleted after a permanently-archived restart file has been produced.
```

atteggiani commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+A dictionary to run scripts or subcommands at various stages of a _payu_ submission:
+
+- `error` gets called if the model does not run correctly and returns an error code.
+- `run` gets called after the model successful execution, but prior to model output archive.
```
Is this after each model run or after the whole experiment (e.g., `payu run -n ...` )? 
```suggestion
- `run` gets called after each the entire model experiment completes successfully, but before the model output is archived.
```

jo-basevi replied:
It's run after the model run command completes e.g. `mpirun ...`. 

Currently there isn't an option to run a script after all of a series of payu runs have completed with `payu run -n ..`.

atteggiani reviewed (changes requested):
Thank you all for your contributions to this page. 

Overall I really like it and I think it explains things clearly. 

I added quite a lot of suggestions, but they are mostly related to formatting or phrasing.

atteggiani reviewed (changes requested):
Thank you all for your contributions to this page. 

Overall I really like it and I think it explains things clearly. 

I added quite a lot of suggestions, but they are mostly related to formatting or phrasing.

ccarouge committed:
(#978) - start of edits from review

jo-basevi commented on docs/models/run_a_model/payu.md:
````diff
@@ -0,0 +1,474 @@
+project: ol01
+```
+
+For model configurations and output to be saved to a `/scratch` storage allocation other than `project` (or your default if `project` is not set) then also set `shortpath` to the desired path. 
````
Yes short path will need be a full path. E.g. `/scratch/PROJECT_CODE`

jo-basevi reviewed (dismissed):
Just had a re-read through and it all looks good to me. No extra comments sorry!

aidanheerdegen commented:
I don't want to delay this any longer than necessary, but @Qian-HuiChen has [recently added an interactive cloning mode](https://github.com/payu-org/payu/pull/703) that is likely to be the most widely used way of cloning a configuration or experiment from GitHub.

It's still being tested, but it's likely we'll release fairly soon as it's a nice feature for users. Consequently I think it should be included in these docs.

ccarouge commented on docs/models/run_a_model/payu.md:
````diff
@@ -0,0 +1,474 @@
+      - '*.nc.*'
+      - 'iceh.????-??-??.nc'
+```
+To enable syncing, change `enable` to `True`, and set `path` to a location on `/g/data`, where _payu_ will copy output and restart folders. A sensible `path` could be: `/g/data/$PROJECT/$USER/<model>/<experiment_name>/`.
````
I see the `basepath` stuff has been released, but can't find it in the docs: https://payu.readthedocs.io/en/stable/search.html?q=basepath

@aidanheerdegen @Qian-HuiChen I'm not reading the PR to understand what it does...

Qian-HuiChen replied:
Hi Claire, thanks for reaching out! Yeah definitely no need to read the PR because we are here 😃 
Instructions about `base_path` are on [payu ReadTheDoc sync section](https://payu.readthedocs.io/en/stable/config.html#:~:text=base_path,will%20be%20overwritten.). Please let me know if you need anything else :)

ccarouge commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+These options are typically useful to modify the physics used in the model, the input data, or the model variables saved in the output files.
+
+These configuration options are specified in files located inside a subfolder of the _control_ directory, named according to the submodel's `name` specified in the `config.yaml` `submodels` section (e.g., configuration options for the _ocean_ component are in the `ocean` sub-directory).<br>
+To modify these options please refer to the configurations documentation of the respective model component, found on the [Run a Model][Run a Model] page for your chosen model.
```
References to what?

ccarouge committed:
(#978) - More changes from review

ccarouge committed:
(#978) - Addressed bigger review comments.

ccarouge commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+These options are typically useful to modify the physics used in the model, the input data, or the model variables saved in the output files.
+
+These configuration options are specified in files located inside a subfolder of the _control_ directory, named according to the submodel's `name` specified in the `config.yaml` `submodels` section (e.g., configuration options for the _ocean_ component are in the `ocean` sub-directory).<br>
+To modify these options please refer to the configurations documentation of the respective model component, found on the [Run a Model][Run a Model] page for your chosen model.
```
I have tried something to move the information of the model components to the structure bit. I've left a shorten version at the bottom of the page as well, for people who skip the structure bit.

ccarouge committed:
(#978) - Add colours to `payu clone` output for readability

Qian-HuiChen commented:
@ccarouge Thanks for adding `payu clone` interactive mode in the documents! It looks very nice 👍 

Thanks for pointing out the inconsistent use of `control directory` and `local experiment directory`. I agree that it could be confusing. I have updated this question to 
```
What would you like to name your local control directory?
```
in commit [271fac5](https://github.com/payu-org/payu/pull/719/commits/271fac5f8ef46d6bab0099c34d0eddfa1af6e712). This will be included in the next payu release :)

aidanheerdegen commented on docs/assets/payu_directory_structure.png:
Is there a source file that can be included as well so the diagram can be modified/updated as required?

ccarouge replied:
Nope, need a Lucid Chart account to access the source.

aidanheerdegen replied:
Doh. Can you export an SVG that could be edited using something like Inkscape?

aidanheerdegen commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,505 @@
+As shown in the diagram, the general layout of a _payu_-supported model run consists of two main directories:
+
+- The _control_ directory contains the model configuration and is the directory from which the model run is started.  
+  {: #control-directory }. This directory contains information to manage the simulation, it also contains the scientific options for each model component that define the physics used in the model component or the diagnostics saved by the model component. These configuration options are specified in files located inside a subfolder of the _control_ directory, named according to the submodel's `nam ...
```
Model sub-directories are only required for coupled models with more than one component. Note that for all the currently released models this is true, and I'm not sure it is worth going into now. However the test AM3 payu config did not have sub-models for example. And neither would stand-along CABLE.

ccarouge replied:
So in these cases, where are the namelists etc. located?

anton-seaice replied:
typically the namelists/config files would be in the model directory, no submodels would be specified in the config.yaml 

([example](https://github.com/ACCESS-NRI/access-om3-configs/blob/dev-MC_25km_jra_iaf/config.yaml))

aidanheerdegen commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+
+- identify the `<repository>` and `<branch>` name the configuration is stored under on GitHub. See the information on the [Run a Model][Run a Model] page of your chosen model for this step.
+- decide where on Gadi to store all your _payu_ experiments, `<configurations-directory>`, typically a folder under $HOME. This directory must exist before running _payu_.
+- decide on a name for your experiment, `<local-branch>`. It is recommended to choose a descriptive name, specific to your experiment.
```
Ok. But I think the name of the experiment and the control-directory should be swapped. People seem to get really confused on this point and name their branch the entire experiment name with the control directory name included. 

It should be stressed that the control directory is the "top level" of the experiment "name-space". All experiments/branches will include the control directory name as a common element.

ccarouge replied:
So you are saying in the example, the branch should simply be `test1` while the control directory is `diffuse_exps-1deg_jra55_ryf`? I guess I need to find a better idea than `test1` as an experiment name for this theoretical diffusion research.

aidanheerdegen commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,505 @@
+_payu_ provides the `payu status` command for monitoring jobs (see [documentation](https://payu.readthedocs.io/en/stable/usage.html#monitoring-payu-jobs)). This command can return the scheduler job ID, and the stage the payu run is currently at. When the job is complete, it displays the exit statuses from the model and overall payu run, and points to the PBS log files. 
+
+!!! note
+    `payu status` is available in _payu_ versions 1.2.0 and later. This command does not yet support monitoring post-processing jobs from the configuration, e.g. `payu collate` and `payu sync`.
```
I think it would be good to have some example `payu status` output. If you agree @ccarouge then @Qian-HuiChen could provide something suitable.

ccarouge replied:
That would be great.

aidanheerdegen replied:
@Qian-HuiChen can you provide some example `payu status` output? It would be fine to paste in a comment as a code block.

Qian-HuiChen replied:
Here is an example of `payu status` output for archived jobs:
```shell
========================================
Run: 8
  Job ID:            archive_example.gadi-pbs
  Run ID:            xxxx
  Stage:             archive
  Total Queue Time:  0h 1m 7s
  Model Finish Time: 1950-10-01T00:00:00
  Exit Status:       0 (Success)
  Model Exit Code:   0 (Success)
  Output Log:        /home/189/USER/expt.o100
  Error Log:         /home/189/USER/expt.3100
  Job File:          /scratch/tm70/USER/archive/expt-branch—6dhash/payu_jobs/8/run/archive_example.gadi-pbs.json
========================================
```
For running jobs, the current experiment time is displayed instead of the model finish time.
```shell
========================================
Run: 8
  Job ID:            running_example.gadi-pbs
  Run ID:            xxxx
  Stage:             model-run
  Current Expt Time: 1950-10-01T00:00:00
  Exit Status:       0 (Success)
  Model Exit Code:   0 (Success)
  Output Log:        /home/189/USER/expt.o100
  Error Log:         /home/189/USER/expt.3100
  Job File:          /scratch/tm70/USER/archive/expt-branch—6dhash/payu_jobs/8/run/running_example.gadi-pbs.json
========================================
```
To monitor the current queue time of a queuing job, please use `payu status --update`.

aidanheerdegen commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,474 @@
+### Start the run from a specific restart file {: id='specific-restart'}
+
+To configure the experiment to start from specific restart files, add a [`restart:` entry](https://payu.readthedocs.io/en/stable/config.html#miscellaneous) to the `config.yaml` file, specifying the path to a folder containing existing restart files.
+Or, to do this automatically when setting up an experiment, add the `-r` flag to the `payu clone` command. 
```
> Or, to do this automatically when setting up an experiment using `payu clone` interactive, give the restart path when prompted `Do you want to specify a custom restart path?`.

aidanheerdegen commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,505 @@
+sync:
+    enable: False # set path below and change to true
+    restart: True
+    path: none # Set to location on /g/data or a remote server and path (rsync syntax)
```
```suggestion
   base_path: none # Set to location on /g/data or a remote server and path (rsync syntax)
```
We should use `base_path` now. See below and https://github.com/payu-org/payu/issues/724#issuecomment-4293083472 for context

aidanheerdegen reviewed (commented):
I have added some more comments. Hope that is helpful.

ccarouge committed:
Use `base_path` instead of `path` for sync option

New `base_path` option is safer since payu will create a uniq folder for the experiment.

ccarouge committed:
Update instructions for custom restart path in payu.md

Clarify instructions for specifying a custom restart path during experiment setup.

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,505 @@
+
+## About
+
+[_payu_](https://github.com/payu-org/payu) is a workflow manager tool for running numerical models in supercomputing environments. It is an open-source software, distributed under an Apache 2.0 Licence.
```
```suggestion
[_payu_](https://github.com/payu-org/payu) is a workflow manager tool for running numerical models in supercomputing environments. It is an open-source software package, distributed under an Apache 2.0 Licence.
```

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,505 @@
+As shown in the diagram, the general layout of a _payu_-supported model run consists of two main directories:
+
+- The _control_ directory contains the model configuration and is the directory from which the model run is started.  
+  {: #control-directory }. This directory contains information to manage the simulation, it also contains the scientific options for each model component that define the physics used in the model component or the diagnostics saved by the model component. These configuration options are specified in files located inside a subfolder of the _control_ directory, named according to the submodel's `nam ...
```
<img width="396" height="91" alt="Image" src="https://github.com/user-attachments/assets/17773914-9364-4c77-adce-5e09db073cd0" />

Something doesn't render right

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,505 @@
+  {: #control-directory }. This directory contains information to manage the simulation, it also contains the scientific options for each model component that define the physics used in the model component or the diagnostics saved by the model component. These configuration options are specified in files located inside a subfolder of the _control_ directory, named according to the submodel's `nam ...
+- The _laboratory_ directory contains all data from _payu_ experiments of the same model. By default, it is `/scratch/$PROJECT/$USER/<model_name>`. `$PROJECT` and `$USER` are environment variables on _Gadi_ that points to your [default project](/getting_started/set_up_nci_account/#change-default-project-on-gadi) and your username respectively. See the section on [modifying the PBS resources](#mod ...
+
+On _Gadi_, it is good practice to put experiment _control_ directories in your `$HOME` directory as this is the only filesystem that is actively backed-up. There is a 10GB limit for home directories, but the _control_ directory only contains text files and symlinks, and so uses relatively little space (<1MB). The _laboratory_ directory is on `/scratch` which is optimised for fast I/O for large da ...
```
```suggestion
On _Gadi_, it is good practice to put experiment _control_ directories in your `$HOME` directory as this is the only filesystem that is actively backed-up. There is a 10GB limit for home directories, but the _control_ directory only contains text files and symlinks, and so uses relatively little space (<1MB). The _laboratory_ directory is on `/scratch` which is optimised for fast IO for large data and where there is adequate space available for large model output.  
```

I/O splits over lines, maybe IO is better?

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,505 @@
+To get a local copy of a configuration, you need to:
+
+- identify the `<repository>` and `<branch>` name the configuration is stored under on GitHub. See the information on the [Run a Model][Run a Model] page of your chosen model for this step.
+- decide where on Gadi to store all your _payu_ experiments, `<configurations-directory>`, typically a folder under $HOME. This directory must exist before running _payu_.
```
```suggestion
- decide where on Gadi to store all your _payu_ experiments, `<configurations-directory>`, typically a folder under `$HOME`. This directory must exist before running _payu_.
```

anton-seaice commented on docs/models/run_a_model/payu.md:
```diff
@@ -0,0 +1,505 @@
+
+#### Submodels {: .no-toc }
+
+Coupled models typically deploy multiple submodels, a.k.a. the model components.
```
```suggestion
Coupled models may deploy the model components as multiple submodels.
```

ccarouge committed:
(#978) - responses to reviews

ccarouge committed:
Replace diagram PNG with SVG
